rust/src/io/commit/dynamodb.rs

```diff
@@ -0,0 +1,492 @@
+// Copyright 2023 Lance Developers.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! DynamoDB based external manifest store
+//!
+
+use std::sync::Arc;
+
+use async_trait::async_trait;
+use aws_sdk_dynamodb::error::SdkError;
+use aws_sdk_dynamodb::operation::get_item::builders::GetItemFluentBuilder;
+use aws_sdk_dynamodb::operation::put_item::builders::PutItemFluentBuilder;
+use aws_sdk_dynamodb::operation::query::builders::QueryFluentBuilder;
+use aws_sdk_dynamodb::types::{AttributeValue, KeyType};
+use aws_sdk_dynamodb::Client;
+use snafu::OptionExt;
+
+use crate::error::{IOSnafu, NotFoundSnafu};
+use crate::io::commit::external_manifest::ExternalManifestStore;
+use crate::{Error, Result};
+
+impl<E> From<SdkError<E>> for Error {
+    fn from(e: SdkError<E>) -> Self {
+        Self::IO {
+            message: format!("dynamodb error: {e}"),
+        }
+    }
+}
+
+#[derive(Debug)]
+struct DynamoDBExternalManifestStore {
+    client: Client,
+    table_name: String,
+}
+
+// these are in macro because I want to use them in a match statement
+macro_rules! base_uri {
+    () => {
```

Comment:
is this necessary? saw the constant value for these two values as well.

Reply:
Yeah, I needed this because I use these in a match statement. If I do
```rust
match (some_name) {
    BASE_URI => ...
}
```
Rust interprets this as trying to assign value in `some_name` into `BASE_URI`, not `if some_name == BASE_URI`.

With a macro I can then do
```rust
match (some_name) {
    base_uri!() => ...
}
```

I guess I can just change every thing to macros instead of static variables